File: lintcode/partition_list/cpp/partition_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/partition-list
@Language: C++
@Datetime: 17-02-15 23:43
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param x: an integer
     * @return: a ListNode 
     */
    ListNode *partition(ListNode *head, int x) 
    {
        auto ltdummy = ListNode{ 0 };
        auto ltnode  = &ltdummy;
        
        auto gtdummy = ListNode{ 0 };
        auto gtnode  = &gtdummy;
        
        auto curr    = head;
        
        while (curr != nullptr)
        {
            if (curr->val < x) {
                ltnode->next = curr;
                ltnode       = ltnode->next;
            } else {
                gtnode->next = curr;
                gtnode       = gtnode->next;
            }
            
            curr = curr->next;
        }
        
        ltnode->next = gtdummy.next;
        gtnode->next = nullptr;
        
        return ltdummy.next;
    }
};

File: lintcode/partition_list/py/partition_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/partition-list
@Language: Python
@Datetime: 17-02-15 23:36
'''

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param x: an integer
    @return: a ListNode 
    """
    def partition(self, head, x):
        ltdummy = ListNode(0)
        ltnode  = ltdummy
        
        gtdummy = ListNode(0)
        gtnode  = gtdummy
        
        curr    = head
        
        while curr != None:
            if curr.val < x:
                ltnode.next = curr
                ltnode      = ltnode.next
            else:
                gtnode.next = curr
                gtnode      = gtnode.next
                
            curr = curr.next
            
        ltnode.next = gtdummy.next
        gtnode.next = None
        
        return ltdummy.next